tape/src/main/java/com/squareup/tape/QueueFile.java

```diff
@@ -456,6 +458,68 @@ public synchronized int forEach(ElementVisitor reader) throws IOException {
     return elementCount;
   }
 
+  @Override public Iterator<byte[]> iterator() {
+    return new ElementIterator();
+  }
+
+  private final class ElementIterator implements Iterator<byte[]> {
+    /** Index of element to be returned by subsequent call to next. */
+    private int cursor = 0;
+
+    /** Position of element to be returned by subsequent call to next. */
+    private int cursorPosition = first.position;
+
+    /**
+     * Size recorded at construction (also in remove) to stop iterator and also to check for
+     * concurrent modification.
+     */
+    private int size = elementCount;
+
+    @Override public boolean hasNext() {
+      return cursor != size;
+    }
+
+    private byte[] readNext() {
+      try {
+        // Read the current element
+        Element current = readElement(cursorPosition);
+        byte[] buffer = new byte[current.length];
+        cursorPosition = wrapPosition(current.position + Element.HEADER_LENGTH);
+        ringRead(cursorPosition, buffer, 0, current.length);
+
+        // Update the pointer to the next element
+        cursorPosition = wrapPosition(current.position + Element.HEADER_LENGTH + current.length);
+
+        // Return the read element
+        return buffer;
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    @Override public byte[] next() {
+      if (cursor == size) throw new NoSuchElementException();
+
+      byte[] element = readNext();
+
+      if (elementCount != size) throw new ConcurrentModificationException();
```

Comment:
I don't think this is enough of a guard against concurrent modification. Equal counts of removal and addition would pass by without exception.